llvm/labs/lab2/kashirin_alexander/BitShiftPass.cpp

```diff
@@ -0,0 +1,116 @@
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/Pass.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Support/CommandLine.h"
+
+static llvm::cl::opt<bool> MulShiftConstOnly(
+    "mul-shift-const-only", llvm::cl::init(false),
+    llvm::cl::desc(
+        "Limit mul-to-shl replacement to constant pow_of_two operands"));
+
+namespace {
+struct BitShiftPass : llvm::PassInfoMixin<BitShiftPass> {
+public:
+  llvm::PreservedAnalyses run(llvm::Function &Func,
+                              llvm::FunctionAnalysisManager &FAM) {
+    std::vector<llvm::Instruction *> toRemove;
+    bool changed = false;
+    for (llvm::BasicBlock &BB : Func) {
+      for (llvm::Instruction &Inst : BB) {
+        if (!llvm::BinaryOperator::classof(&Inst)) {
+          continue;
+        }
+        llvm::BinaryOperator *op = llvm::cast<llvm::BinaryOperator>(&Inst);
+        if (op->getOpcode() != llvm::Instruction::BinaryOps::Mul) {
+          continue;
+        }
+
+        llvm::Value *leftOper = op->getOperand(0);
+        llvm::Value *rightOper = op->getOperand(1);
+
+        int logVal1 = getLogBase2(leftOper);
+        int logVal2 = getLogBase2(rightOper);
```

Comment:
What do `logVar1` and `logVar2` mean?

Reply:
logVar1 is a base 2 logarithm of left operand, logVar2 is a base 2 logarithm of right operand